### ISSUE TITLE
Validate buffer dtype in pre-forward hook for FSDP mixed precision tests

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#104740]
### 🚀 The feature, motivation and pitch

As per the comment in https://github.com/pytorch/pytorch/pull/104682/files, `test_full_precision_in_eval_buffers` would be more robust if we used an nn.Module pre-forward hook to validate the expected buffer dtype, instead of doing so after the forward pass. This would ensure that the computation took place in the right dtype.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

**Comment by awgu** [comment]
I am self-assigning to claim as a ramp-up task.

**Comment by awgu** [comment]
The unit test to improve:
https://github.com/pytorch/pytorch/blob/71d18f61059dbb617a8afc79b64598d2fbb89d63/test/distributed/fsdp/test_fsdp_mixed_precision.py#L860

**Background**
- An [`nn.Module`](https://pytorch.org/docs/stable/generated/torch.nn.Module.html) inductively defines a logical model and its layers. A module can be in training mode (`self.training == True`) or eval mode (`self.training == False`), which can be configured via `module.train()` and `module.eval()`, respectively.
- The PyTorch fully sharded data parallel (FSDP) implementation shards module parameters ([`nn.Parameter`](https://pytorch.org/docs/stable/generated/torch.nn.parameter.Parameter.html)s) but not buffers ([`register_buffer()`](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_buffer)).
- However, FSDP still manages buffers to some degree, e.g. it offers a [mixed precision API](https://github.com/pytorch/pytorch/blob/b65b9e6ff430137f5d6ddca5017a00d7dc7f3564/torch/distributed/fsdp/api.py#L130-L135) where buffers can be cast to a low precision for computation.
- We have an environment variable `FSDP_USE_FULL_PREC_IN_EVAL` that if set to `1`, forces FSDP-managed parameters and buffers to full precision (fp32 / the `torch.float32` dtype) if the modules are in eval mode. If set to `0`, then those parameters/buffers are kept in low precision for computation.
- The unit test checks if that the buffers are in the expected dtype for a few different settings (`use_composable`, `cast_forward_inputs`, and the aforementioned env var). (The expected dtype depends on the setting.)

**Overview**
The unit test currently checks that the buffers are in the expected precision by checking the model's buffer dtypes after the forward pass finishes.
https://github.com/pytorch/pytorch/blob/71d18f61059dbb617a8afc79b64598d2fbb89d63/test/distributed/fsdp/test_fsdp_mixed_precision.py#L890-L894

In theory, we could pass this unit test even if we cast the buffer dtypes _after_ the forward computation has finished, while we care that the buffer dtype is already the expected one for the computation.

Thus, for this task, we want to register a module forward pre-hook on each module in the model that has a buffer, where that hook checks that the buffer is in the expected dtype.
- To register a module forward pre-hook, use [`nn.Module.register_forward_pre_hook(hook, ...)`](https://github.com/pytorch/pytorch/blob/71d18f61059dbb617a8afc79b64598d2fbb89d63/test/distributed/fsdp/test_fsdp_mixed_precision.py#L890-L894).
- To know which modules to register the hook on, you can check something like:
```
if len(list(module.buffers(recurse=False))) > 0:
    module.register_forward_pre_hook(hook, ...)
```
- For the hook itself, you can iterate over the module's buffers and reuse the assertion `self.assertEqual(expected_dtype, buf.dtype)`.

**Comment by awgu** [comment]
https://github.com/pytorch/pytorch/pull/106231
